code/scripts/DataTableGen.hs

```diff
@@ -0,0 +1,292 @@
+config configFilePath = do
+  setCurrentDirectory configFilePath
+  c <- readFile "DTG_Config.txt"
+  let l = map words $ filter isInfoLine (lines c)
+      -- FIXME: use real parser to extract settings from config file
+      -- will improve error messages regarding config file settings
+      (packageNames,classInstOrd) = (head l, map toClassType (l !! 1))
+  return (packageNames,classInstOrd)
+
+-- function creates and writes output data file DataTable.csv to /code/analysis
+output :: FilePath -> EntryString -> [ClassName] -> IO ()
+output outputFilePath entryData ordClassInsts = do
+  createDirectoryIfMissing False outputFilePath
+  setCurrentDirectory outputFilePath
+  dataTable <- openFile "DataTable.csv" WriteMode
+  hPutStrLn dataTable "Package,\t,\t,\t,\t,\t,Class Instances"
+  hPutStr dataTable "drasil-,File Path,File Name,Data Type,Newtype Type,Class Definitions,"
+  hPutStrLn dataTable (intercalate "," ordClassInsts)
+  hPutStrLn dataTable entryData
+  hClose dataTable
+
+-- creates an entry for each file (new Entry data-oriented format)
+createEntry :: FilePath -> DC.File -> DC.FileName -> IO Entry
+-- creates blank line to separate file entries by drasil- package
+createEntry homeDirectory file "newline" = return nlEntry where 
+  nlEntry = makeEntry "" "newline" "" [] [] [] []
+-- creates actual file entries
+createEntry homeDirectory file filename = do
+  let drpk = DC.fileDrasilPack file
+      fn = filename
+      fp = DC.filePath file
+      -- entry file path is stripped of home directory and drasil- pack
+      efp = (++ "/") $ (\\ homeDirectory ++ "/drasil-" ++ drpk ++ "/") fp
+
+  -- extracts entry data from File data type
+  -- stripInstances = [(dataType,classInfo)]
+  (dtl,ntl,classNames,stripInstances) <- SCR.extractEntryData fn fp
+  
+  let clstp
+        | drpk == "gool" = GOOL
+        | otherwise = Drasil
+
+  let cls = map (makeClass clstp) classNames
+      clsint = map makeClassInstance stripInstances
+
+  let entry = makeEntry drpk fn efp dtl ntl cls clsint
+  return entry
+
+-- creates entrystring for each entry (contains lines for each entry's data)
+compileEntryData :: [ClassName] -> Entry -> DC.FileName -> IO EntryString
+-- creates blank line entrystring to separate file entries by drasil- package
+compileEntryData ordClassInsts entry "newline" = return "\t"
+-- creates entrystrings for actual file entries
+compileEntryData ordClassInsts entry filename = do
+
+  -- joins data, newtype and class values/placeholders (first data entry line)
+  let f d n c = intercalate "," [drpk,fp,fn,d,n,c]
+      (drpk,fp,fn) = (drasilPack entry,filePath entry,fileName entry)
+
+  -- extracts data, newtype and class names + class instances from entry
+  let dataNames = dataTypes entry
+      newtypeNames = newtypes entry
+      classNames = map className (classes entry)
+      entryClassInsts = classInstances entry
+
+  -- guards determine how to handle first data entry line
+  let entry 
```

Comment:
You should use pattern-matching here instead of `not null`. `head` is always a sign that you're doing something odd.

Reply:
By pattern matching, I assume you mean something like the following:
```haskell
entry (x:_) = f x "\t" "\t"
```
But I am unsure by what you mean by using pattern-matching instead of `not null`; this was included since `(x:_)` can't be patterned-matched to `[]` (it raises a non-exhaustive pattern exception). What pattern-matching work-around are you looking for the replace the use of `null null`?

Reply:
Yes, that's what I mean by pattern matching.  Matching on `(x:_)` does insure the list is non-empty, which is what `not null` means in this case.

Reply:
I'm unsure as to how I can pattern-match `(x:_)` onto an explicit list like `dataNames` for this. Does it happen to involve the use of `@` at all?

Reply:
Just to be sure, is it better to use `where` for this (like shown below),
```haskell
let entry = hEnt dataNames newtypeNames classNames where
      hEnt (x:_) _ _ = f x "\t" "\t"
      hEnt _ (x:_) _ = f "\t" x "\t"
      hEnt _ _ (x:_) = f "\t" "\t" x
      hEnt _ _ _     = f "\t" "\t" "\t"
```
or not (like shown below)?
```haskell
let entry = hEnt dataNames newtypeNames classNames
    hEnt (x:_) _ _ = f x "\t" "\t"
    hEnt _ (x:_) _ = f "\t" x "\t"
    hEnt _ _ (x:_) = f "\t" "\t" x
    hEnt _ _ _     = f "\t" "\t" "\t"
```

Reply:
I refractored the definition for `entry` [here](https://github.com/JacquesCarette/Drasil/pull/2208/commits/4065330bf5885d36437fb97d322a5c23d339df61#) (old lines 179-183; new lines 178-182).